pkg/agent/controller/networkpolicy/reconciler.go

```diff
@@ -541,18 +542,27 @@ func (r *reconciler) computeOFRulesForAdd(rule *CompletedRule, ofPriority *uint1
 		return ofRuleByServicesMap, lastRealized
 	}
 	if rule.Direction == v1beta2.DirectionIn {
+		isRuleAppliedToService := isRuleAppliedToService(rule.TargetMembers)
 		// Addresses got from source GroupMembers' IPs.
 		from1 := groupMembersToOFAddresses(rule.FromAddresses)
 		// Get addresses that in From IPBlock but not in Except IPBlocks.
-		from2 := ipBlocksToOFAddresses(rule.From.IPBlocks, r.ipv4Enabled, r.ipv6Enabled)
+		from2 := ipBlocksToOFAddresses(rule.From.IPBlocks, r.ipv4Enabled, r.ipv6Enabled, isRuleAppliedToService)
 		membersByServicesMap, servicesMap := groupMembersByServices(rule.Services, rule.TargetMembers)
 		for svcKey, members := range membersByServicesMap {
-			ofPorts := r.getOFPorts(members)
-			lastRealized.podOFPorts[svcKey] = ofPorts
+			var toAddresses []types.Address
+			if isRuleAppliedToService {
+				svcGroupIDs := r.getSvcGroupIDs(members)
+				toAddresses = svcGroupIDsToOFAddresses(svcGroupIDs)
+				lastRealized.serviceGroupIDs = svcGroupIDs
```

Comment:
I just found this implicitly asserts there is only one item in `membersByServicesMap`, otherwise the assignment would override previous write, could we add a comment for it?

Reply:
Added.